// demux-js/dist/demux/readers/testing/JsonActionReader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractActionReader_1 = require("../AbstractActionReader");
const request_promise_native_1 = __importDefault(require("request-promise-native"));
/**
 * Reads from an array of `Block` objects, useful for testing.
 */
class JsonActionReader extends AbstractActionReader_1.AbstractActionReader {
    constructor(blockchain, startAtBlock = 1, onlyIrreversible = false, maxHistoryLength = 600, requestInstance = request_promise_native_1.default) {
        super(startAtBlock, onlyIrreversible, maxHistoryLength);
        this.blockchain = blockchain;
        this.startAtBlock = startAtBlock;
        this.onlyIrreversible = onlyIrreversible;
        this.maxHistoryLength = maxHistoryLength;
        this.requestInstance = requestInstance;
    }
    getHeadBlockNumber() {
        return __awaiter(this, void 0, void 0, function* () {
            const block = this.blockchain.slice(-1)[0];
            if (this.blockchain.length !== block.blockNumber) {
                throw Error(`Block at position ${this.blockchain.length} indicates position ${block.blockNumber} incorrectly.`);
            }
            return block.blockNumber;
        });
    }
    getBlock(blockNumber) {
        return __awaiter(this, void 0, void 0, function* () {
            const block = this.blockchain[blockNumber - 1];
            if (!block) {
                throw Error(`Block at position ${blockNumber} does not exist.`);
            }
            if (block.blockNumber !== blockNumber) {
                throw Error(`Block at position ${blockNumber} indicates position ${block.blockNumber} incorrectly.`);
            }
            return block;
        });
    }
}
exports.JsonActionReader = JsonActionReader;
