// demux-js/dist/demux/watchers/BaseActionWatcher.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Cooredinates implementations of `AbstractActionReader`s and `AbstractActionHandler`s in
 * a polling loop.
 */
class BaseActionWatcher {
    constructor(actionReader, actionHandler, pollInterval) {
        this.actionReader = actionReader;
        this.actionHandler = actionHandler;
        this.pollInterval = pollInterval;
    }
    /**
     * Starts a polling loop running in replay mode.
     */
    replay() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.actionReader.seekToBlock(this.actionReader.startAtBlock);
            yield this.watch();
        });
    }
    /**
     * Uses the given actionReader and actionHandler to poll and process new blocks.
     */
    watch() {
        return __awaiter(this, void 0, void 0, function* () {
            // Record start time
            const startTime = new Date().getTime();
            // Process blocks until we're at the head block
            let headBlockNumber = 0;
            while (!headBlockNumber || this.actionReader.currentBlockNumber < headBlockNumber) {
                const [blockData, isRollback] = yield this.actionReader.nextBlock();
                // Handle block (and the actions within them)
                let needToSeek = false;
                let seekBlockNum = 0;
                if (blockData) {
                    [needToSeek, seekBlockNum] = yield this.actionHandler.handleBlock(blockData, isRollback, this.actionReader.isFirstBlock);
                }
                // Seek to next needed block at the request of the action handler
                if (needToSeek) {
                    yield this.actionReader.seekToBlock(seekBlockNum - 1);
                }
                headBlockNumber = this.actionReader.headBlockNumber;
            }
            // Record end time
            const endTime = new Date().getTime();
            // Calculate timing for next iteration
            const duration = endTime - startTime;
            let waitTime = this.pollInterval - duration;
            if (waitTime < 0) {
                waitTime = 0;
            }
            // Schedule next iteration
            setTimeout(() => __awaiter(this, void 0, void 0, function* () { return yield this.watch(); }), waitTime);
        });
    }
}
exports.BaseActionWatcher = BaseActionWatcher;
